File: modules/hydromate-voice/android/src/main/java/expo/modules/hydromatevoice/HydroMateVoiceModule.kt
package expo.modules.hydromatevoice

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

internal data class VoicePreferencesConfiguration(
  @Field val enabled: Boolean = false,
  @Field val language: String = "en",
  @Field val selectedVoiceIdentifier: String? = null,
  @Field val waterSpeechTemplate: String = "",
  @Field val medicineSpeechTemplate: String = "",
  @Field val lockedMedicineSpeechTemplate: String = "",
  @Field val birthdaySpeechTemplate: String = "",
  @Field val lockedBirthdaySpeechTemplate: String = "",
  @Field val anniversarySpeechTemplate: String = "",
  @Field val lockedAnniversarySpeechTemplate: String = "",
  @Field val customSpeechTemplate: String = "",
  @Field val lockedCustomSpeechTemplate: String = "",
  @Field val localeCandidates: List<String> = emptyList()
) : Record

class HydroMateVoiceModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("HydroMateVoice")

    AsyncFunction("scheduleVoiceTest") { secondsFromNow: Int ->
      val context = appContext.reactContext?.applicationContext
        ?: throw IllegalStateException("Android application context is unavailable")

      VoiceAlarmScheduler.scheduleTest(context, secondsFromNow).toDouble()
    }

    AsyncFunction("scheduleWaterVoiceAlarm") {
      identifier: String,
      expectedAtMillis: Double,
      amountMl: Double,
      repeatsDaily: Boolean ->
      val context = applicationContext()
      VoiceAlarmScheduler.scheduleWater(
        context,
        identifier,
        expectedAtMillis.toLong(),
        amountMl,
        repeatsDaily
      ).toDouble()
    }

    AsyncFunction("cancelWaterVoiceAlarm") { identifier: String ->
      VoiceAlarmScheduler.cancelWater(applicationContext(), identifier)
    }

    AsyncFunction("cancelAllWaterVoiceAlarms") {
      VoiceAlarmScheduler.cancelAllWater(applicationContext())
    }

    AsyncFunction("scheduleMedicineVoiceAlarm") {
      identifier: String,
      expectedAtMillis: Double,
      medicineName: String,
      repeatsDaily: Boolean ->
      VoiceAlarmScheduler.scheduleMedicine(
        applicationContext(),
        identifier,
        expectedAtMillis.toLong(),
        medicineName,
        repeatsDaily
      ).toDouble()
    }

    AsyncFunction("cancelMedicineVoiceAlarm") { identifier: String ->
      VoiceAlarmScheduler.cancelMedicine(applicationContext(), identifier)
    }

    AsyncFunction("cancelAllMedicineVoiceAlarms") {
      VoiceAlarmScheduler.cancelAllMedicine(applicationContext())
    }

    AsyncFunction("scheduleDatedVoiceAlarm") {
      identifier: String,
      expectedAtMillis: Double,
      reminderType: String,
      reminderText: String,
      repeatsYearly: Boolean ->
      VoiceAlarmScheduler.scheduleDated(
        applicationContext(),
        identifier,
        expectedAtMillis.toLong(),
        reminderType,
        reminderText,
        repeatsYearly
      ).toDouble()
    }

    AsyncFunction("cancelDatedVoiceAlarm") {
      identifier: String,
      reminderType: String ->
      VoiceAlarmScheduler.cancelDated(
        applicationContext(),
        identifier,
        reminderType
      )
    }

    AsyncFunction("cancelAllDatedVoiceAlarms") { reminderType: String ->
      VoiceAlarmScheduler.cancelAllDated(applicationContext(), reminderType)
    }

    AsyncFunction("syncVoicePreferences") {
      configuration: VoicePreferencesConfiguration ->
      NativeVoicePreferences.sync(
        applicationContext(),
        configuration.enabled,
        configuration.language,
        configuration.selectedVoiceIdentifier,
        configuration.waterSpeechTemplate,
        configuration.medicineSpeechTemplate,
        configuration.lockedMedicineSpeechTemplate,
        configuration.birthdaySpeechTemplate,
        configuration.lockedBirthdaySpeechTemplate,
        configuration.anniversarySpeechTemplate,
        configuration.lockedAnniversarySpeechTemplate,
        configuration.customSpeechTemplate,
        configuration.lockedCustomSpeechTemplate,
        configuration.localeCandidates
      )
    }

    AsyncFunction("setVoiceRemindersEnabled") { enabled: Boolean ->
      NativeVoicePreferences.setEnabled(applicationContext(), enabled)
    }

    AsyncFunction("setSelectedVoiceIdentifier") {
      language: String,
      identifier: String? ->
      NativeVoicePreferences.setSelectedVoice(
        applicationContext(),
        language,
        identifier
      )
    }
  }

  private fun applicationContext() =
    appContext.reactContext?.applicationContext
      ?: throw IllegalStateException("Android application context is unavailable")
}

File: modules/hydromate-voice/src/HydroMateVoiceModule.ts
import { requireOptionalNativeModule } from "expo-modules-core";
import { Platform } from "react-native";

type HydroMateVoiceNativeModule = {
  scheduleVoiceTest: (secondsFromNow: number) => Promise<number>;
  scheduleWaterVoiceAlarm?: (
    identifier: string,
    expectedAtMillis: number,
    amountMl: number,
    repeatsDaily: boolean
  ) => Promise<number>;
  cancelWaterVoiceAlarm?: (identifier: string) => Promise<void>;
  cancelAllWaterVoiceAlarms?: () => Promise<void>;
  scheduleMedicineVoiceAlarm?: (
    identifier: string,
    expectedAtMillis: number,
    medicineName: string,
    repeatsDaily: boolean
  ) => Promise<number>;
  cancelMedicineVoiceAlarm?: (identifier: string) => Promise<void>;
  cancelAllMedicineVoiceAlarms?: () => Promise<void>;
  scheduleDatedVoiceAlarm?: (
    identifier: string,
    expectedAtMillis: number,
    reminderType: DatedVoiceReminderType,
    reminderText: string,
    repeatsYearly: boolean
  ) => Promise<number>;
  cancelDatedVoiceAlarm?: (
    identifier: string,
    reminderType: DatedVoiceReminderType
  ) => Promise<void>;
  cancelAllDatedVoiceAlarms?: (
    reminderType: DatedVoiceReminderType
  ) => Promise<void>;
  syncVoicePreferences?: (
    configuration: NativeVoicePreferencesConfiguration
  ) => Promise<void>;
  setVoiceRemindersEnabled?: (enabled: boolean) => Promise<void>;
  setSelectedVoiceIdentifier?: (
    language: string,
    identifier: string | null
  ) => Promise<void>;
};

export type DatedVoiceReminderType = "birthday" | "anniversary" | "custom";

export type NativeVoicePreferencesConfiguration = {
  enabled: boolean;
  language: string;
  selectedVoiceIdentifier: string | null;
  waterSpeechTemplate: string;
  medicineSpeechTemplate: string;
  lockedMedicineSpeechTemplate: string;
  birthdaySpeechTemplate: string;
  lockedBirthdaySpeechTemplate: string;
  anniversarySpeechTemplate: string;
  lockedAnniversarySpeechTemplate: string;
  customSpeechTemplate: string;
  lockedCustomSpeechTemplate: string;
  localeCandidates: string[];
};

const nativeModule =
  requireOptionalNativeModule<HydroMateVoiceNativeModule>("HydroMateVoice");

export const isLockedVoiceTestAvailable =
  Platform.OS === "android" && nativeModule !== null;

export const isWaterVoiceCompanionAvailable =
  Platform.OS === "android" &&
  typeof nativeModule?.scheduleWaterVoiceAlarm === "function";

export const isMedicineVoiceCompanionAvailable =
  Platform.OS === "android" &&
  typeof nativeModule?.scheduleMedicineVoiceAlarm === "function";

export const isDatedVoiceCompanionAvailable =
  Platform.OS === "android" &&
  typeof nativeModule?.scheduleDatedVoiceAlarm === "function";

export async function scheduleVoiceTest(secondsFromNow: number) {
  if (!nativeModule) {
    throw new Error(
      "The HydroMate native voice module is not included in this build."
    );
  }

  return nativeModule.scheduleVoiceTest(secondsFromNow);
}

export async function scheduleWaterVoiceAlarm(
  identifier: string,
  expectedAtMillis: number,
  amountMl: number,
  repeatsDaily = true
) {
  if (!nativeModule?.scheduleWaterVoiceAlarm) {
    return false;
  }

  await nativeModule.scheduleWaterVoiceAlarm(
    identifier,
    expectedAtMillis,
    amountMl,
    repeatsDaily
  );
  return true;
}

export async function cancelWaterVoiceAlarm(identifier: string) {
  await nativeModule?.cancelWaterVoiceAlarm?.(identifier);
}

export async function cancelAllWaterVoiceAlarms() {
  await nativeModule?.cancelAllWaterVoiceAlarms?.();
}

export async function scheduleMedicineVoiceAlarm(
  identifier: string,
  expectedAtMillis: number,
  medicineName: string,
  repeatsDaily = true
) {
  if (!nativeModule?.scheduleMedicineVoiceAlarm) {
    return false;
  }

  await nativeModule.scheduleMedicineVoiceAlarm(
    identifier,
    expectedAtMillis,
    medicineName,
    repeatsDaily
  );
  return true;
}

export async function cancelMedicineVoiceAlarm(identifier: string) {
  await nativeModule?.cancelMedicineVoiceAlarm?.(identifier);
}

export async function cancelAllMedicineVoiceAlarms() {
  await nativeModule?.cancelAllMedicineVoiceAlarms?.();
}

export async function scheduleDatedVoiceAlarm(
  identifier: string,
  expectedAtMillis: number,
  reminderType: DatedVoiceReminderType,
  reminderText: string,
  repeatsYearly = true
) {
  if (!nativeModule?.scheduleDatedVoiceAlarm) {
    return false;
  }

  await nativeModule.scheduleDatedVoiceAlarm(
    identifier,
    expectedAtMillis,
    reminderType,
    reminderText,
    repeatsYearly
  );
  return true;
}

export async function cancelDatedVoiceAlarm(
  identifier: string,
  reminderType: DatedVoiceReminderType
) {
  await nativeModule?.cancelDatedVoiceAlarm?.(identifier, reminderType);
}

export async function cancelAllDatedVoiceAlarms(
  reminderType: DatedVoiceReminderType
) {
  await nativeModule?.cancelAllDatedVoiceAlarms?.(reminderType);
}

export async function syncNativeVoicePreferences(
  configuration: NativeVoicePreferencesConfiguration
) {
  if (!nativeModule?.syncVoicePreferences) {
    return;
  }

  await nativeModule.syncVoicePreferences(configuration);
}

export async function setNativeVoiceRemindersEnabled(enabled: boolean) {
  await nativeModule?.setVoiceRemindersEnabled?.(enabled);
}

export async function setNativeSelectedVoiceIdentifier(
  language: string,
  identifier: string | null
) {
  await nativeModule?.setSelectedVoiceIdentifier?.(language, identifier);
}
